src/lib/Tisch/Run.hs

```diff
@@ -172,27 +172,27 @@ withoutPerm
   :: (MonadIO m, Cx.MonadMask m, Allow p ps, ps' ~ DropPerm p ps)
   => proxy (p :: k)
   -- ^ @k@ may be 'Perm' or @['Perm']@.
-  -> Conn ps
-  -> (Conn ps' -> m a)
+  -> Conn ps d
+  -> (Conn ps' d -> m a)
   -- ^ The usage of @'Conn' ps@ is undefined within this function,
   -- and @'Conn' ps'@ mustn't escape the scope of this function.
   -> m a
 withoutPerm _ (Conn conn) f = f (Conn conn)
 
 -- | Open a new connection.
-connect :: MonadIO m => Pg.ConnectInfo -> m Conn'
-connect = connect' . Pg.postgreSQLConnectionString
+connect :: MonadIO m => proxy d -> Pg.ConnectInfo -> m (Conn' d)
+connect pd info = connect' pd $ Pg.postgreSQLConnectionString info
 
 -- | Like 'connect', except it takes a @libpq@ connection string instead of a
 -- 'Pg.ConnectInfo'.
-connect' :: MonadIO m => B8.ByteString -> m Conn'
-connect' = liftIO . fmap Conn . Pg.connectPostgreSQL
+connect' :: MonadIO m => proxy d -> B8.ByteString -> m (Conn' d)
```

Comment:
We can do without the `proxy` argument here and in `connect'`. The rationale is that the usage we give to this connection will fix to the desired type anyway, so type inference is not particularly hindered. And, if by some reason we want to explicitly type `d` at `connect'`s call site, then we can do so with an explicit type signature.